### PR TITLE
Token signing key is now randomly generated on demand if not already present

### PR DESCRIPTION
**Pull request by dseevr**
Additionally, ensure that the PRNG has been seeded

Signed-off-by: Bill Robinson <dseevr@users.noreply.github.com>

### CODE REVIEW

File: auth/token.go [hunk 1]
```diff
@@ -19,9 +23,12 @@ const (
 	// TokenValidityInHours represents the token validity; used to set token expiry
 	TokenValidityInHours = 10
 
-	// TokenSigningKey is used for signing the token
-	// FIXME: this should be fetched from store
-	TokenSigningKey = "ccn$!~56"
+	// this is the datastore key, not the actual bytes used for signing.
+	// since changes are not handled by an abstraction layer, construct the path manually
+	tokenSigningKey = types.AuthProxyDir + "/token_signing_key"
```
Comment (yuva29) [inline auth/token.go]: This the place we maintain all the datastore paths https://github.com/contiv/auth_proxy/blob/master/db/constants.go. And, we use `GetPath` to get the absolute datastore path like here https://github.com/contiv/auth_proxy/blob/master/db/local.go#L61

I recommend to follow the same style.

File: auth/token.go [hunk 2]
```diff
@@ -30,6 +37,70 @@ const (
...
 )
 
+func init() {
+	// ensures we don't generate predictable token signing keys
+	rand.Seed(time.Now().UnixNano())
+}
+
+var tokenCharacters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890")
+
+func generateTokenSigningKey() string {
```
Comment (yuva29) [inline auth/token.go]: What do you think of following approach 3 which seems to be fast http://stackoverflow.com/questions/22892120/how-to-generate-a-random-string-of-a-fixed-length-in-golang

And, this func can be named `generateRandString`. Yay??

File: auth/token.go [hunk 3]
```diff
@@ -30,6 +37,70 @@ const (
...
+	stateDrv, err := state.GetStateDriver()
+	if err != nil {
+		return "", err
+	}
+
+	signingKeyMutex.Lock()
+	defer signingKeyMutex.Unlock()
+
+	// check for an existing existing key
+	existingKey, err := stateDrv.Read(tokenSigningKey)
```
Comment (yuva29) [inline auth/token.go]: I feel this is more clear this way  
```goexistingKey, err := stateDrv.Read(tokenSigningKey)
 switch(err) {
 case err == auth_errors.ErrKeyNotFound:
 case err == auth_errors.ErrKeyExists:
 default:
  return "", err
 
 }

File: common/crypto.go [hunk 4]
```diff
@@ -24,6 +27,10 @@ const (
 	cost = 13
 )
 
+func init() {
```
Comment (yuva29) [inline common/crypto.go]: Why is this here?

Comment (dseevr) [inline common/crypto.go]: This package was calling crypto-related stuff without a guaranteed-to-be-seeded PRNG

File: auth/token.go (hunk 3 above)
Comment (dseevr) [inline auth/token.go]: Done

File: auth/token.go (hunk 2 above)
Comment (dseevr) [inline auth/token.go]: I don't think there's any value in extracting this into a reusable function right now.  The valid characters are also specific to the token signing key.

This function is being called extremely infrequently.  The cumulative CPU time of everyone who will ever have this code execute on their systems would never add up to the time it would take to replace it with an optimized version.  The third option there also doesn't have the same probability for all characters to be chosen.

File: auth/token.go (hunk 1 above)
Comment (dseevr) [inline auth/token.go]: Done

**Comment by dseevr** [comment]
PTAL again

**Comment by yuva29** [comment]
Please add test so we can be sure that this works fine.

File: systemtests/auth_test.go [hunk 5]
```diff
@@ -4,12 +4,66 @@ import (
...
+	c.Assert(err, IsNil)
+	secondSigningKey := string(data)
+
+	// make sure the token works
+	resp, _ = proxyGet(c, secondToken, endpoint)
+	c.Assert(resp.StatusCode, Equals, 200)
+
+	// make sure the old token does not work
+	resp, _ = proxyGet(c, firstToken, endpoint)
+	c.Assert(resp.StatusCode, Equals, 400)
```
Comment (yuva29) [inline systemtests/auth_test.go]: Please check the body to be `Bad Token`

Comment (dseevr) [inline systemtests/auth_test.go]: Done, but this didn't get collapsed because I didn't edit this line
